[Su Peigeng] iP

Bacon-Strips opened the pull request:
## Duke
> Task managing chat-bot

Duke is able to help you remember things, it is 

- Text based
- User friendly
- Fast to use

To download, all you have to do is
1. Download the JAR file [here](https://github.com/Bacon-Strips/ip/releases/tag/A-Jar)
2. Open your PowerShell
3. Navigate the PowerShell to the folder where Duke,jar is in
4. Type in the PowerShell `java -jar Duke.jar`
5. Enjoy 😉 

Furthermore, it is **FREE**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (Coming soon)
- [ ] Reminders (Coming soon)
---
This is a java project made for CS2103T. Duke can be started in other `main` function this way
```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Bacon-Strips committed:
Working skeletal functions for the chatbot

Bacon-Strips committed:
Removed debugging expressions and the logo static attribute

Bacon-Strips committed:
Level 2 features, including a list of tasks that Duke can save

Bacon-Strips committed:
Added the Task class to the chatbot

Bacon-Strips committed:
Added the functionality to mark and unmark tasks within the chatbot

Bacon-Strips committed:
Added 3 new classes TodoTask, DeadlineTask and EventTask

Bacon-Strips committed:
Editted the input splitter function to accomodate todo, event, deadine

Bacon-Strips committed:
Implemented the deadine, todo and event tasks into Duke

Bacon-Strips committed:
Now Duke will track the amount of tasks you have as you add in new
tasks.

Bacon-Strips committed:
Implemented the UI testing component and modified some of the inputSplit
function such that errorenous inputs will not be passed into the tasks.

Bacon-Strips committed:
Added the DukeException Exception and modified Duke to make use of it.

Bacon-Strips committed:
Previous commit accidentally commited an incomplete implementation of
DukeException. This one is complete.

Also, delete functionality added.

Bacon-Strips committed:
Add a Map containing constructors for tasks

Addition of Map consolidates constructors for various tasks into a
single data structure, reducing repeated codes within Duke.

Initially, Duke has many repeated codes for the various cases depending
on what is the task called.

Bacon-Strips committed:
Add Map taskOperations

This map contains the functions that specify how to deal with the
various user commands within Duke.

The code is initially lengthened by the use of if else statements to
check for various commands. Now they are replaced with a single
encompassing if statement that retrieves the required function to handle
user request.

Bacon-Strips committed:
Fix bugs

Bacon-Strips committed:
Add functionality to retrieve task data

Now Duke is able to retrieve data from a text file saved within the data
folder.

Bacon-Strips committed:
Add function to write tasks to disk

Now Duke is able to write the tasklist to disk, allowing future
rebooting of Duke to access the same tasklist

Bacon-Strips committed:
Add LocalDate support for Deadline and Event tasks

Bacon-Strips committed:
Merge branches 'branch-Level-7' and 'branch-Level-8'

Merge Level 7 and 8 features

Bacon-Strips committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Bacon-Strips committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Bacon-Strips committed:
Modify TaskMaker to make Tasks using String array

Bacon-Strips committed:
Add Ui class

Bacon-Strips committed:
Add Storage class

Bacon-Strips committed:
Add TaskList class

Bacon-Strips committed:
Modify Storage class to save to disk

Bacon-Strips committed:
Add Command class

Bacon-Strips committed:
Update Duke to split work to many classes

Bacon-Strips committed:
Merge branch 'master' of github.com:Bacon-Strips/ip

Bacon-Strips committed:
Add java files to a single duke package

Bacon-Strips committed:
Reallocate classes among different packages

Bacon-Strips committed:
Implement some unit testing for Storage class

Bacon-Strips committed:
Repackage java files

Bacon-Strips committed:
Document all public classes and methods

Bacon-Strips committed:
Update JavaDocs

Bacon-Strips committed:
Add FindCommand class

This commit adds the FindCommand class, which allows Duke to find tasks
with matching key words.

A search function within the chat-bot is more user-friendly when there
are many tasks within the TaskList.

Yuvaraj0702 commented on .gitignore:
```diff
@@ -13,5 +13,9 @@ src/main/resources/docs/
 *.iml
 bin/
 
+*.class
+data/tasks.txt
+testDir/testfile.txt
+
```
Good job this is a good and easy way to omit your files.

Yuvaraj0702 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,77 @@
+ * </ul>>
+ *
+ * @author Su Peigeng
+ */
```
Well written JavaDoc

Yuvaraj0702 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,77 @@
+ * @author Su Peigeng
+ */
+public class Duke {
+    private final TaskList tasks;
+    private final Storage storage;
+    private final Ui ui;
```
Constants should be in all caps. 
```suggestion
    private final TaskList TASKS;
    private final Storage STORAGE;
    private final Ui UI;
```

Yuvaraj0702 commented on src/main/java/seedu/duke/command/Command.java:
```diff
@@ -0,0 +1,33 @@
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
+
+    /**
+     * Indicates if the Command is the exit command.
+     *
+     * @return      true only and only if is exit command
+     */
+    public boolean isExit() {
+        return false;
+    }
+}
```
Feel like this function is unnecessary. You could just use the Boolean false instead.

Yuvaraj0702 commented on src/main/java/seedu/duke/operations/Parser.java:
```diff
@@ -0,0 +1,47 @@
+     * @return      Command
+     */
+    public static Command parse(String cmd) {
+        cmd = cmd.strip();
```
Preferably do not use short forms.
```suggestion
    public static Command parse(String command) {
        command = command.strip();
```

Yuvaraj0702 commented on src/main/java/seedu/duke/operations/Storage.java:
```diff
@@ -0,0 +1,143 @@
+        File dataFile = new File(filePath);
+        if (!dataFile.exists()) {
+            setupDirectory();
+            return new ArrayList<>();
+        } else {
+            ArrayList<Task> loadedTasks = new ArrayList<>();
```
Although type inference is valid it would be better to explicitly state types.
```suggestion
            return new ArrayList<Task>();
        } else {
            ArrayList<Task> loadedTasks = new ArrayList<Task>();
```

Yuvaraj0702 reviewed (commented):
Great job with your code. You managed to do the optional tasks which was really commendable. Your code in general seems very well written and adheres to the code quality guidelines. There are some violations and suggestions I have given as well to improve general readability of the code. Great job and keep it up. :grinning:

Bacon-Strips committed:
Update src/main/java/seedu/duke/Duke.java

Co-authored-by: Yuvaraj0702 <89921842+Yuvaraj0702@users.noreply.github.com>

Bacon-Strips committed:
Update src/main/java/seedu/duke/operations/Parser.java

Co-authored-by: Yuvaraj0702 <89921842+Yuvaraj0702@users.noreply.github.com>

Bacon-Strips committed:
Add Java Launcher JavaFX support

Bacon-Strips committed:
Add echo functionality to Duke Launcher

Bacon-Strips committed:
Add DialogBox

Bacon-Strips committed:
Implement GUI for Duke

Bacon-Strips committed:
Merge branch 'branch-Level-10'

Bacon-Strips committed:
Merge branch 'master' of github.com:Bacon-Strips/ip

Bacon-Strips committed:
Fix bugs

Bacon-Strips committed:
Add assertions to various methods

There are several methods that requires important assumptions about the
input. In some cases if the assumption does not hold the program will
run into errors.

Adding assertions highlight important assumptions made.

Assertions help verify that the assumptions made holds true within the
methods. Doing so will make debugging easier.

Bacon-Strips committed:
Improve code quality

Code quality for project improved in accordance to Java standards and
conventions.

Coding in accordance to Java standards and conventions makes the project
more readable to other developers to work on and debug.

Let's edit the package names, variable names and comments such that they
are now in accordance to convention.

Bacon-Strips committed:
Merge pull request #2 from Bacon-Strips/branch-A-Assertions

Add assertions to various methods

Bacon-Strips committed:
Merge branch 'master' into branch-A-CodeQuality

Bacon-Strips committed:
Merge pull request #3 from Bacon-Strips/branch-A-CodeQuality

Improve code quality

Bacon-Strips committed:
Add functionality to detect clashing Event Tasks

Bacon-Strips committed:
Merge branch 'branch-B-DetectAnomalies'

Bacon-Strips committed:
Improve GUI and UX for Duke

Bacon-Strips committed:
Improve GUI

Bacon-Strips committed:
Implement application shutdown procedure

Bacon-Strips committed:
Change location of README.md fil

Bacon-Strips committed:
Bug fix

Bacon-Strips committed:
Add Ui.png

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Edit README.md

Bacon-Strips committed:
Remove redundant shadowJar files

Bacon-Strips committed:
Merge branch 'master' of github.com:Bacon-Strips/ip
